ui/components/app/contact-list/contact-list.test.js

```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import { shallowWithContext } from '../../../../test/lib/render-helpers';
+import RecipientGroup from './recipient-group';
+import ContactList from '.';
+
+describe('Contact List', () => {
+  describe('given searchForContacts', () => {
+    it('sorts contacts by name within each letter group', () => {
+      const contacts = {
+        Al: { name: 'Al', address: '0x0' },
+        aa: { name: 'aa', address: '0x1' },
+        Az: { name: 'Az', address: '0x2' },
+        Bl: { name: 'Bl', address: '0x3' },
+        ba: { name: 'ba', address: '0x4' },
+        Bz: { name: 'Bz', address: '0x5' },
+        Ccc: { name: 'Ccc', address: '0x6' },
+      };
+      const searchForContacts = () => {
+        return Object.values(contacts);
+      };
+      const selectRecipient = () => null;
+      const selectedAddress = null;
+
+      const wrapper = shallowWithContext(
+        <ContactList
+          searchForContacts={searchForContacts}
+          selectRecipient={selectRecipient}
+          selectedAddress={selectedAddress}
+        />,
+      );
+
+      expect(wrapper).toMatchElement(
```

Comment:
I opted to use `jest-enzyme`, which provides this matcher, but happy to use something else if we prefer another way.

Reply:
Okay, I think I know why we don't use this library. It seems to have a dependency on an old version of `jest-environment-jsdom` which is majorly screwing up tests. So, I think this isn't a great idea.

Reply:
Okay, I opted to copy this matcher from Enzyme (see comment above for reasoning).

Reply:
Great work here! I'm generally in favor of using something like react-test-library more for UI testing but I'm honestly not sure if it can do the check for element ordering in the DOM that we're looking for here... 🤔

Reply:
Yeah! It seems that the JS community has largely embraced react-testing-library over Enyzme, and it seems that react-testing-library has some more TLC around it than Enzyme, so I am not sure whether we are wanting to move in that direction or not? Checking the entire content of a component rather than its pieces doesn't seem to be something that's common for react-testing-library, but perhaps we could make use of [`toContainHTML`](https://github.com/testing-library/jest-dom#tocontainhtml)?

Reply:
I feel like in this particular case react-testing-library won't work. I wouldn't block this PR on this. Though there might be folks on the team who don't want to introduce new uses of enzyme. Not sure what teamwide feelings are on this. cc @danjm @darkwing @Gudahtt @NiranjanaBinoy @ryanml

Reply:
I'm not sure I follow why `react-testing-library` would be ill-suited for this purpose :thinking: . It has an extensive querying API that should allow grabbing the list contents and checking the order.

Reply:
>Checking the entire content of a component rather than its pieces doesn't seem to be something that's common for react-testing-library, but perhaps we could make use of toContainHTML?

"Checking the entire contents of a component" sounds like a snapshot to me, which is supported by `react-testing-library`. I don't think it has anything like `toMatchElement` though, as you're doing here. I don't think snapshots are appropriate to use here either, because a snapshot would be more difficult to read.

As far as I understand it, `react-testing-library` doesn't support this style of testing _on purpose_. They want you to test for the things users see, e.g. ensure this text is present, in this order. The focus is on the expected output, not in the implementation details. Whereas this test, as written, looks for a `RecipientComponent` with a specific set of inputs. If the inputs were renamed or the contents of the component restructured, this test might break even under conditions where the list is still sorted alphanumerically, making it fragile.

Reply:
Cool, thanks for those pointers. For some reason I didn't notice we were already using react-testing-library. I reworked the test to use it instead of a custom matcher. Let me know if that's better!

Reply:
Nicely done! It looks like there are a bunch of methods that I either just didn't know about or have been added to `react-testing-library` since I last wrote a true UI test 🤦 I think what you have now looks alot better!